tests/unit/autogram/test_algo_3.py

```diff
@@ -0,0 +1,140 @@
+    # Re-compute the activations because autogram_ needs the non-freed graph
+    activations = compute_activations(criterion, input, model, target)
+    _ = autogram_(activations, batch_size, criterion, model)
+    activations = compute_activations(criterion, input, model, target)
+    with Timer(device=DEVICE):
+        gramian = autogram_(activations, batch_size, criterion, model)
+
+    activations = compute_activations(criterion, input, model, target)
+    _ = compute_gramian_via_autojac(activations, model)
+    activations = compute_activations(criterion, input, model, target)
+    with Timer(device=DEVICE):
+        expected_gramian = compute_gramian_via_autojac(activations, model)
+
+    assert_close(gramian, expected_gramian)
+
+
+class Timer:
+    def __init__(self, device=None):
+        self.device = device
+        self.start_time = None
+        self.end_time = None
+
+    def __enter__(self):
+        if self.device == "cuda":
+            torch.cuda.synchronize()
+        self.start_time = time.perf_counter()
+        return self  # This allows you to access the timer object within the 'with' block
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.device == "cuda":
+            torch.cuda.synchronize()
+        self.end_time = time.perf_counter()
+        self.elapsed_time = self.end_time - self.start_time
+        print(f"Elapsed time: {self.elapsed_time:.4f} seconds")
+
+
+def compute_gramian_via_autojac(activations, model: nn.Sequential):
+    params = OrderedSet(model.parameters())
+    output = OrderedSet([activations[-1]])
+    init = Init(output)
+    diag = Diagonalize(output)
+    jac = Jac(output, params, chunk_size=1)
+    matrixify = _Matrixify()
+    transform = matrixify << jac << diag << init
+    jacobian_matrices = transform({})
+    expected_gramian = torch.stack([J @ J.T for J in jacobian_matrices.values()]).sum(dim=0)
+    return expected_gramian
+
+
+def compute_activations(criterion, input, model: nn.Sequential, target) -> list[Tensor]:
+    activations = [input]
+    for layer in model:
+        activation = layer(activations[-1])
+        activations.append(activation)
+
+    losses = criterion(activations[-1], target)
+    activations.append(losses)
+    return activations
+
+
+def autogram_(activations, batch_size, criterion, model: nn.Sequential):
+    grad = torch.ones_like(activations[-1])
+    gramian = torch.zeros(batch_size, batch_size)
+    for i, (input, output, layer) in list(
+        enumerate(zip(activations[:-1], activations[1:], list(model) + [criterion]))
+    )[::-1]:
+        params = list(layer.parameters())
+        if len(params) > 0:
+
+            def get_vjp(input_j, grad_output_j) -> tuple[Tensor, ...]:
+                return vjp_from_module(layer, input_j)(grad_output_j)
+
+            jacobians = torch.vmap(get_vjp)(input, grad)
+
+            assert len(jacobians) == 1
```

Comment:
This is a problem right? we are supposed to have as many jacobians as we have parameters?